### ISSUE TITLE
OneCycleLR updating on epoch rather than on step

### ISSUE THREAD

**Issue by grudloff** [pytorch/pytorch#65417]
## 🐛 Bug

OneCycleLR makes a change to learning rate (and momentum) supposedly on every step, but it seems that instead, it is doing an update on every epoch, which leads to unwanted behavior since only a fraction of the schedule is followed. As of now, the solution is to set `total_steps=epochs` but it has been said that a change for every batch within an epoch is an important part of this schedule.

## To Reproduce

```python
# setup model & dataloaders
steps_per_epoch = len(train_dataloader) 
epochs = epochs# any number
# add optimizer
lr_scheduler = torch.optim.lr_scheduler.OneCycleLR(optimizer, 
                                      max_lr=1e-1,
                                      steps_per_epoch=steps_per_epoch,
                                      epochs=epochs, verbose=True)
#setup trainer & fit
```

This code will lead to only `epochs` updates of lr&momentum and to change it only to a fraction of `steps_per_epoch/epochs` of the intended schedule. 

Example output from Boring model with `steps_per_epoch=1000 `and `epochs=5`
```
Adjusting learning rate of group 0 to 4.0001e-03.
Adjusting learning rate of group 0 to 4.0004e-03.
Adjusting learning rate of group 0 to 4.0009e-03.
Adjusting learning rate of group 0 to 4.0017e-03.
Adjusting learning rate of group 0 to 4.0026e-03.
```
As can be seen, only a small fraction of the up step of the schedule was performed, not achieving lr even near to 1e-1.

[TheBoringModel](https://gist.github.com/grudloff/e787a627a0f39a240d1955a74f3232d6)

**Comment by grudloff** [comment]
Thanks a lot @cowwoc, it is working now! I am a newcomer to the pytorch/pytorch-lightning and I was unknowingly facing this issue for some time, I can only imagine the number of people falling into this issue! I wouldn't have guessed this was a pytorch-lightning issue, but now it is crystal clear.

As a newcomer, it seemed obvious to me that just passing a scheduler was enough for it to work as expected and that a config dictionary was just necessary for more granular control. I think the most sensible thing would be to add a warning in the documentation about this or add an example of the usefulness of the config dictionary when a scheduler uses a non-default (per-step) updating scheme.

I don't think there is much to do from pytorch's side. I think this deserves an issue on pytorch-lightning's github. I am going to take a look at the related issues there and then post an issue there.

**Comment by soulitzer** [comment]
Closing since this appears to be a PyTorch lightning issue
